src/runtime/security/kinit_context.cpp

```diff
@@ -0,0 +1,306 @@
+#include <dsn/utility/time_utils.h>
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/utility/flags.h>
+#include <dsn/utility/filesystem.h>
+#include <dsn/utility/smart_pointers.h>
+
+namespace dsn {
+namespace security {
+extern bool FLAGS_enable_auth;
+
+#define KRB5_RETURN_NOT_OK(err, msg)                                                               \
+    do {                                                                                           \
+        krb5_error_code __err_code__ = (err);                                                      \
+        if (__err_code__ != 0) {                                                                   \
+            return krb5_call_to_errors(__err_code__, (msg));                                       \
+        }                                                                                          \
+    } while (0);
+
+DSN_DEFINE_string("security", krb5_keytab, "", "absolute path of keytab file");
+DSN_DEFINE_string("security", krb5_config, "", "absolute path of krb5_config file");
+DSN_DEFINE_string("security", krb5_principal, "", "kerberos principal");
+
+// Attention: we can't do these check work by `DSN_DEFINE_validator`, because somebody may don't
+// want to use security, so these configuration may not setted. In this situation, these checks
+// will not pass.
+error_s check_configuration()
+{
+    dassert(FLAGS_enable_auth,
+            "There is no need to check configuration if FLAGS_enable_auth is not true");
+
+    if (0 == strlen(FLAGS_krb5_keytab) || !utils::filesystem::file_exists(FLAGS_krb5_keytab)) {
+        return error_s::make(ERR_INVALID_PARAMETERS,
+                             fmt::format("invalid keytab file \"{}\"", FLAGS_krb5_keytab));
+    }
+
+    if (0 == strlen(FLAGS_krb5_config) || !utils::filesystem::file_exists(FLAGS_krb5_config)) {
+        return error_s::make(ERR_INVALID_PARAMETERS,
+                             fmt::format("invalid krb5 config file \"{}\"", FLAGS_krb5_config));
+    }
+
+    if (0 == strlen(FLAGS_krb5_principal)) {
+        return error_s::make(ERR_INVALID_PARAMETERS, "empty principal");
+    }
+
+    return error_s::ok();
+}
+
+class kinit_context_impl
+{
+public:
+    kinit_context_impl() : _opt(nullptr) {}
+    virtual ~kinit_context_impl();
+    // implementation of 'kinit -k -t <keytab_file> <principal>'
+    error_s kinit();
+
+private:
+    // init kerberos context
+    void init_krb5_ctx();
+
+    // get _user_name from _principal
+    error_s parse_username_from_principal();
+
+    // get or renew credentials from KDC and store it to _ccache
+    error_s get_credentials();
+    void schedule_renew_credentials();
+
+    error_s wrap_krb5_err(krb5_error_code krb5_err, const std::string &msg);
+    error_s krb5_call_to_errors(krb5_error_code krb5_code, const std::string &prefix_msg);
+
+private:
+    krb5_context _krb5_context;
+    // krb5 principal
+    krb5_principal _principal;
+    // keytab file with absolute path
```

Comment:
This comment makes confused.

Reply:
What makes you confused? 
Sorry I don't know how to explain this to you. Because it's the basic conceptive in kerberos.